File: RandomForest.py
import pandas as pd
import numpy as np

wbcd = pd.read_csv("C:\\Datasets_BA\\Python Scripts\\wbcd.csv")

wbcd = wbcd.iloc[:,1:32] # Excluding id column

# converting B to Benign and M to Malignant 
wbcd['diagnosis'] = np.where(wbcd['diagnosis'] == 'B','Benign ',wbcd['diagnosis'])
wbcd['diagnosis'] = np.where(wbcd['diagnosis'] == 'M','Malignant ',wbcd['diagnosis'])

# Splitting data into training and testing data set
from sklearn.model_selection import train_test_split
train,test = train_test_split(wbcd,test_size = 0.2)

X_train = np.array(train.ix[:,1:31]) # Input
X_test = np.array(test.ix[:,1:31]) # Input
Y_train = np.array(train['diagnosis']) # Output
Y_test = np.array(test['diagnosis']) # Output

# Normalization function 
def norm_func(i):
    x = (i-i.min())	/ (i.max()-i.min())
    return (x)

# Normalized data frame (considering the numerical part of data)
train_n = norm_func(train.iloc[:,1:])
train_n.describe()

test_n = norm_func(test.iloc[:,1:])
test_n.describe()

from sklearn.ensemble import RandomForestClassifier

rfmodel = RandomForestClassifier(n_estimators=15)

rfmodel.fit(X_train,Y_train)


# Train Data Accuracy
train["rf_pred"] = rfmodel.predict(X_train)
train_acc = np.mean(train["diagnosis"]==train["rf_pred"])
train_acc

# Test Data Accuracy
test["rf_pred"] = rfmodel.predict(X_test)
test_acc = np.mean(test["diagnosis"]==test["rf_pred"])
test_acc 

File: RandomForest.R
############ WBCD #################
# Read the dataset
wbcd <- read.csv(file.choose())

#First colum in dataset is id which is not required so we will be taking out
wbcd <- wbcd[-1]

#table of diagonis B <- 357 and M <- 212
table(wbcd$diagnosis)

# Replace B with Benign and M with Malignant. Diagnosis is factor with 2 levels that is B and M. We also replacing these two entery with Benign and Malignat
wbcd$diagnosis <- factor(wbcd$diagnosis, levels = c("B","M"), labels = c("Benign","Malignant"))

# table or proportation of enteries in the datasets. What % of entry is Bengin and % of entry is Malignant
round(prop.table(table(wbcd$diagnosis))*100,1)

#Create a function to normalize the data
norm <- function(x){ 
     return((x-min(x))/(max(x)-min(x)))
}

#Apply the normalization function to wbcd dataset
wbcd_n <- as.data.frame(lapply(wbcd[2:31], norm))

# create training and test data
wbcd_train <- wbcd_n[1:469, ]
wbcd_test <- wbcd_n[470:569, ]

# create labels for training and test data
wbcd_train_labels <- wbcd[1:469, 1]
wbcd_test_labels <- wbcd[470:569, 1]


# Building a random forest model on training data 
install.packages("randomForest")
library(randomForest)

wbcd_forest <- randomForest(wbcd_train_labels~.,data=wbcd_train,importance=TRUE)
plot(wbcd_forest)

# Train Data Accuracy
train_acc <- mean(wbcd_train_labels==predict(wbcd_forest))
train_acc

# Test Data Accuracy
test_acc <- mean(wbcd_test_labels==predict(wbcd_forest, newdata=wbcd_test))
test_acc

varImpPlot(wbcd_forest)
